aep/general/0146/aep.md.j2

```diff
@@ -1,5 +1,97 @@
+specify one of the fields inside the `oneof`. Additionally, a `oneof` **may**
+be used with the same type (usually strings) to represent a semantic difference
+between the options.
+
+Because the individual fields in the `oneof` have different keys, a developer
+can programmatically determine which (if any) of the fields is populated.
+
+A `oneof` preserves the largest degree of type safety and semantic meaning for
+each option, and APIs **should** generally prefer them over other generic or
+polymorphic options when feasible. However, the `oneof` construct is ill-suited
+when there is a large (or unlimited) number of potential options, or when there
+is a large resource structure that would require a long series of "cascading
+oneofs".
+
+**Note:** Adding additional possible fields to an existing `oneof` is a
+non-breaking change, but moving existing fields into or out of a `oneof` is
+breaking (it creates a backwards-incompatible change in Go protobuf stubs).
+
+#### Maps
+
+Maps **may** be used in situations where many values _of the same type_ are
+needed, but the keys are unknown or user-determined.
+
+Maps are usually not appropriate for generic fields because the map values all
+share a type, but occasionally they are useful. In particular, a map can
+sometimes be suited to a situation where many objects of the same type are
+needed, with different behavior based on the names of their keys (for example,
+using keys as environment names).
+
+#### Struct
+
+The [`google.protobuf.Struct`][struct] object **may** be used to represent
+arbitrary nested JSON. Keys can be strings, and values can be floats, strings,
+booleans, arrays, or additional nested structs, allowing for an arbitrarily
+nested structure that can be represented as JSON (and is automatically
+represented as JSON when using REST/JSON).
+
+A `Struct` is most useful when the API does not know the schema in advance, or
+when a API needs to store and retrieve arbitrary but structured user data.
+Using a `Struct` is convenient for users in this case because they can easily
+get JSON objects that can be natively manipulated in their environment of
+choice.
+
+If a API needs to reason about the _schema_ of a `Struct`, it **should** use
+[JSONSchema][] for this purpose. Because JSONSchema is itself JSON, a valid
+JSONSchema document can itself be stored in a `Struct`.
+
+#### Any
+
+The [`google.protobuf.Any`][any] object can be used to send an arbitrary
+serialized protocol buffer and a type definition.
+
+However, this introduces complexity, because an `Any` becomes useless for any
+task other than blind data propagation if the consumer does not have access to
+the proto. Additionally, even if the consumer _does_ have the proto, the
+consumer has to ensure the type is registered and then deserialize manually,
+which is an often-unfamiliar process.
+
+Because of this, `Any` **should not** be used unless other options are
```

Comment:
Upgrade this to must not?

Reply:
Yeah the phrasing is kind of funny here, because "unless other options are infeasible" should already be _implied_ by `should not`, right?

I think "**must not**...unless other options are infeasible" is approximately equivalent to "**should not**", so I don't want think we should reword to that.

But "`Any` **should not** be used." also feels a bit terse.

WDYT?

Reply:
Discussed live with @toumorokoshi and concluded this is okay - Alex, please reopen if you disagree.